agent/engine/task_manager_test.go

```diff
@@ -2227,3 +2231,157 @@ func TestTaskWaitForHostResources(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	mtask := &managedTask{
+		Task: &apitask.Task{
+			ResourcesMapUnsafe:  make(map[string][]taskresource.TaskResource),
+			DesiredStatusUnsafe: apitaskstatus.TaskRunning,
+			Volumes: []apitask.TaskVolume{
+				{
+					Name: taskresourcevolume.TestVolumeName,
+					Type: apiresource.EBSTaskAttach,
+					Volume: &taskresourcevolume.EBSTaskVolumeConfig{
+						VolumeId:             taskresourcevolume.TestVolumeId,
+						VolumeName:           taskresourcevolume.TestVolumeId,
+						VolumeSizeGib:        taskresourcevolume.TestVolumeSizeGib,
+						SourceVolumeHostPath: taskresourcevolume.TestSourceVolumeHostPath,
+						DeviceName:           taskresourcevolume.TestDeviceName,
+						FileSystem:           taskresourcevolume.TestFileSystem,
+					},
+				},
+			},
+		},
+		ctx:                      ctx,
+		resourceStateChangeEvent: make(chan resourceStateChange),
+	}
+
+	timeoutctx, timeoutcancel := context.WithTimeout(ctx, 50*time.Millisecond)
+	defer timeoutcancel()
+
+	mockCsiClient := mock_csiclient.NewMockCSIClient(mockCtrl)
+	mockCsiClient.EXPECT().NodeUnstageVolume(gomock.Any(), "vol-12345", "/mnt/ecs/ebs/taskarn_vol-12345").Do(func(ctx context.Context, volID, hostPath string) {
+		time.Sleep(5 * time.Millisecond)
+	}).Return(nil).MinTimes(1)
+
+	errors := mtask.UnstageVolumes(mockCsiClient)
+
+	require.NoError(t, timeoutctx.Err())
+	assert.Len(t, errors, 0)
+}
+
+func TestUnstageVolumesTimeoutError(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	mtask := &managedTask{
+		Task: &apitask.Task{
+			ResourcesMapUnsafe:  make(map[string][]taskresource.TaskResource),
+			DesiredStatusUnsafe: apitaskstatus.TaskRunning,
+			Volumes: []apitask.TaskVolume{
+				{
+					Name: taskresourcevolume.TestVolumeName,
+					Type: apiresource.EBSTaskAttach,
+					Volume: &taskresourcevolume.EBSTaskVolumeConfig{
+						VolumeId:             taskresourcevolume.TestVolumeId,
+						VolumeName:           taskresourcevolume.TestVolumeId,
+						VolumeSizeGib:        taskresourcevolume.TestVolumeSizeGib,
+						SourceVolumeHostPath: taskresourcevolume.TestSourceVolumeHostPath,
+						DeviceName:           taskresourcevolume.TestDeviceName,
+						FileSystem:           taskresourcevolume.TestFileSystem,
+					},
+				},
+			},
+		},
+		ctx:                      ctx,
+		resourceStateChangeEvent: make(chan resourceStateChange),
+	}
+
+	timeoutctx, timeoutcancel := context.WithTimeout(ctx, 5*time.Millisecond)
+	defer timeoutcancel()
+
+	mockCsiClient := mock_csiclient.NewMockCSIClient(mockCtrl)
+	mockCsiClient.EXPECT().NodeUnstageVolume(gomock.Any(), "vol-12345", "/mnt/ecs/ebs/taskarn_vol-12345").Do(func(ctx context.Context, volID, hostPath string) {
+		time.Sleep(50 * time.Millisecond)
```

Comment:
Why is this sleep needed?

Reply:
the sleep is needed here to try to mock a NodeUnstageVolume call taking longer than the timeout duration.

Reply:
I think it's the `Return(fmt.Errorf("rpc error: code = DeadlineExceeded desc = context deadline exceeded")).MinTimes(1)` that's making `NodeStageVolume` return that error and not this sleep. Timeout duration used by NodeUnstageVolume seems to be hardcoded to 3 seconds which is much larger than 50ms sleep that we have here. 
https://github.com/aws/amazon-ecs-agent/blob/f1bea74554c7f2e5ab08c8fc0436c7a0d7a3ff25/agent/engine/task_manager.go#L63C2-L63C22